### ISSUE TITLE
Fancy Indexing for `DeviceNDArray`

### ISSUE THREAD

**Issue by quasiben** [numba/numba#4035]
On the [gitter chat today](https://gitter.im/numba/numba?at=5cc7359b3b6cb0686a74463d) we briefly discussed whether or not to add fancy indexing to Numba back CUDA arrays (DeviceNDArray).  I am interested in this topic as it relates to recent work around cuml and sklearn's gridsearch.  This relevant for context as running `DeviceNDArray` through sklearn machinery means handling code like: `X[indices]`

This issue should serve as a continuation of the discussion on gitter to specifically address the following use case:

```python
import numpy as np
import numba.cuda

data = np.arange(10)
carr = numba.cuda.to_device(data)

carr[data[:5]]
# carr[carr[:5]] # this too ?

# and multidimensional cases

data = np.ones((2,2))
carr = numba.cuda.to_device(data)
carr[np.array([0,1])]
```

cc @cjnolet @seibert @kkraus14

**Comment by seibert** [comment]
To summarize the points from gitter, with some additional context:

* The Numba device array is intentionally very bare-bones GPU array container.  On the host side, the API basically only handles data motion to/from GPU, and creating zero-copy simple slices.  On the device side, we are mainly concerned with element-wise access.

* Host-side array operations that require actual compute, like fancy indexing, are not currently supported.  Originally this was just due to lack of time / requests, but now that CuPy is becoming more popular, we're encouraging people to consider it for these use cases.

* CuPy provides a fully featured GPU array with the NumPy API, including things like fancy indexing.  Numba-compiled GPU functions can take CuPy arrays as arguments, thanks to the `__cuda_array_interface__`.

* That said, CuPy is a pretty heavy dependency if all you are looking for is the scatter functionality of NumPy indexing with integer or boolean arrays.

Writing a CUDA Python kernel to do fancy indexing is mostly straightforward in 1D, but more complicated in N-dimensions.  We could certainly include it in Numba, though I would have the following concerns:
* Is this the "thin end of the wedge" on the road toward recreating a lot of NumPy behavior for Numba device arrays?  Is this sufficiently useful, or are there more operations that will be needed?
* Are these functions run on their own (with the usual kernel launch overhead) fast enough to be useful, or do they need to be fused with other operations in practice?

**Comment by mrocklin** [comment]
+1 on deferring this functionality to cupy (or other numpy-like GPU libraries)

**Comment by seibert** [comment]
Yeah, I still would like to see `DeviceNDArray` slowly evaporate away to be replaced by an external CUDA runtime package for Python (for the barebones memory handle and memory manager) + CuPy/PyTorch/etc (for the full numpy-like experience).
